zkp/circuits/lib/check-nullifier-value.circom

```diff
@@ -0,0 +1,176 @@
+  var isPositive[numOutputs];
+  for (var i = 0; i < numOutputs; i++) {
+    positive[i] = GreaterEqThan(40);
+    positive[i].in[0] <== outputValues[i];
+    positive[i].in[1] <== 0;
+    isPositive[i] = positive[i].out;
+    assert(isPositive[i] == 1);
+  }
+
+  // derive the sender's public key from the secret input
+  // for the sender's private key. This step demonstrates
+  // the sender really owns the private key for the input
+  // UTXOs
+  var inputOwnerPublicKey[2];
+  component pub = BabyPbk();
+  pub.in <== inputOwnerPrivateKey;
+  inputOwnerPublicKey[0] = pub.Ax;
+  inputOwnerPublicKey[1] = pub.Ay;
+
+  // hash the input values
+  component inputHashes[numInputs];
+  var calculatedInputHashes[numInputs];
+  for (var i = 0; i < numInputs; i++) {
+    // perform the hash calculation even though they are not needed when the input 
+    // commitment at the current index is 0; this is because in zkp circuits we
+    // must always perform the same computation (have the the same constraints)
+    inputHashes[i] = Poseidon(4);
+    inputHashes[i].inputs[0] <== inputValues[i];
+    inputHashes[i].inputs[1] <== inputSalts[i];
+    inputHashes[i].inputs[2] <== inputOwnerPublicKey[0];
+    inputHashes[i].inputs[3] <== inputOwnerPublicKey[1];
+    if (inputCommitments[i] == 0) {
+      calculatedInputHashes[i] = 0;
+    } else {
+      calculatedInputHashes[i] = inputHashes[i].out;
+    }
+  }
+
+  // check that the input commitments match the calculated hashes
+  for (var i = 0; i < numInputs; i++) {
+    assert(inputCommitments[i] == calculatedInputHashes[i]);
+  }
+
+  // calculate the nullifier values from the input values
+  component nullifierHashes[numInputs];
+  var calculatedNullifierHashes[numInputs];
+  for (var i = 0; i < numInputs; i++) {
+    nullifierHashes[i] = Poseidon(3);
+    nullifierHashes[i].inputs[0] <== inputValues[i];
+    nullifierHashes[i].inputs[1] <== inputSalts[i];
+    nullifierHashes[i].inputs[2] <== inputOwnerPrivateKey;
+    if (nullifiers[i] == 0) {
+      calculatedNullifierHashes[i] = 0;
+    } else {
+      calculatedNullifierHashes[i] = nullifierHashes[i].out;
+    }
+  }
+
+  // check that the nullifiers match the calculated hashes
+  for (var i = 0; i < numInputs; i++) {
+    assert(nullifiers[i] == calculatedNullifierHashes[i]);
+  }
+
+  // With the above steps, we demonstrated that the nullifiers
+  // are securely bound to the input commitments. Now we need to
+  // demonstrate that the input commitments belong to the Sparse
+  // Merkle Tree with the root `root`.
+  component smtVerifier[numInputs];
+  for (var i = 0; i < numInputs; i++) {
+    smtVerifier[i] = SMTVerifier(nSMTLevels);
+    smtVerifier[i].enabled <== enabled[i];
+    smtVerifier[i].root <== root;
+    for (var j = 0; j < nSMTLevels; j++) {
+      smtVerifier[i].siblings[j] <== merkleProof[i][j];
+    }
+    smtVerifier[i].oldKey <== 0;
+    smtVerifier[i].oldValue <== 0;
+    smtVerifier[i].isOld0 <== 0;
```

Comment:
this is so confusing...

Reply:
will add inline comments based on https://github.com/iden3/circomlib/blob/master/circuits/smt/smtverifier.circom